### PR TITLE
Refactor/sleep and sleep test

### LINKED ISSUE #7: Refactor Sleep Get Average Data
As a dev

When we are trying to **get average sleep data** in multiple methods

I would like to have a **helper function** to run it

so that I can use one method **getAverageOfValues** to get average of any data I pass as arguments

### PR DESCRIPTION
**Pull request by cameronRomo**
## What is the change?
* Refactor for `Sleep.js` and `sleep-test.js`

 ## What does it fix? 
* Code now has better SRP and fewer redundancies
* Test file now has tests for untested methods and has tests for new methods

## Is this a fix or a feature?  
* Fix

## Where should the reviewer start?
**`src/Sleep.js`:**
* line 8

**`test/sleep-test.js`:**
* line 287

 ## How should this be tested?
* Pull down and run `npm test`, then note passing tests under the describe for "sleep"

### CODE REVIEW

File: src/Sleep.js [hunk 1]
```diff
@@ -4,78 +4,104 @@ class Sleep {
...
 
-    let bestSleeperIds = bestSleepers.map(function(bestSleeper) {
+  getSleepersIds(bestSleepers) {
+    return bestSleepers.map(bestSleeper => {
       return (Object.keys(bestSleeper));
     });
+  }
 
-    return bestSleeperIds.map(function(sleepNumber) {
+  getWinnerNamesFromList(sortedArray, userRepo) {
```
Comment (mdflynn) [inline src/Sleep.js:98]: This was a gnarly function. Nice work pulling it out into three separate methods.

**Review by mdflynn — APPROVED** [review]
This was a tough file to read at first, but it's now much easier to follow how the class was designed.

File: src/Sleep.js [hunk 2]
```diff
@@ -4,78 +4,104 @@ class Sleep {
   constructor(sleepData) {
     this.sleepData = sleepData;
   }
+
+  getAverageOfValues(data, sleepProperty = null) {
```
Comment (okalimbet) [inline src/Sleep.js:8]: good reuse of function from the previous classes:)

File: src/Sleep.js [hunk 3]
```diff
@@ -4,78 +4,104 @@ class Sleep {
...
   calculateAverageSleepQuality(id) {
-    let perDaySleepQuality = this.sleepData.filter((data) => id === data.userID);
-    return perDaySleepQuality.reduce((sumSoFar, data) => {
-      return sumSoFar += data.sleepQuality;
-    }, 0) / perDaySleepQuality.length;
+    let perDaySleepQuality = this.filterSleepDataPerDay(id);
+    return this.getAverageOfValues(perDaySleepQuality, "sleepQuality");
+  }
+
+  findSleepDataByDate(id, date) {
```
Comment (okalimbet) [inline src/Sleep.js:28]: Nice replacement for repetitive data!

File: src/Sleep.js [hunk 4]
```diff
@@ -4,78 +4,104 @@ class Sleep {
...
 
+  determineSleepWinnerForWeek(date, userRepo) {
+    let sleepRankWithData = this.getUserSleepObject(date, userRepo, "sleepQuality");
     return this.getWinnerNamesFromList(sleepRankWithData, userRepo);
   }
-  determineSleepHoursWinnerForDay(date, userRepo) {
-    let timeline = userRepo.chooseDayDataForAllUsers(this.sleepData, date);
-    let sleepRankWithData = userRepo.combineRankedUserIDsAndAveragedData(this.sleepData, date, 'hoursSlept', timeline);
 
+  determineSleepHoursWinnerForDay(date, userRepo) {
```
Comment (okalimbet) [inline src/Sleep.js:80]: Great split the old function. It played too many different roles before, it makes more sense to have a separate method for each variable.

**Review by okalimbet — APPROVED** [review]

**Pull request merged**